test_report/src/report.rs

```diff
@@ -0,0 +1,268 @@
+    }
+
+    // saves to local fs and returns the path
+    pub fn save(&self) -> Result<String, anyhow::Error> {
+        let buf = self.serialize_test_result();
+        if let Ok(named_temp_file) = tempfile::Builder::new().suffix(".bin").tempfile() {
+            std::fs::write(&named_temp_file, buf).unwrap_or_default();
+            let (_, path) = named_temp_file.keep().unwrap();
+            let path_str = path.to_str().unwrap_or_default();
+            return Ok(path_str.to_string());
+        }
+        Err(anyhow::anyhow!("Error saving test report"))
+    }
+
+    // adds a test to the test report
+    pub fn add_test(
+        &self,
+        id: Option<String>,
+        name: String,
+        classname: String,
+        file: String,
+        parent_name: String,
+        line: Option<i32>,
+        status: Status,
+        attempt_number: i32,
+        started_at: i64,
+        finished_at: i64,
+        output: String,
+    ) {
+        let mut test = TestCaseRun::default();
+        if let Some(id) = id {
+            test.id = id;
+        }
+        test.name = name;
+        test.classname = classname;
+        test.file = file;
+        test.parent_name = parent_name;
+        if let Some(line) = line {
+            test.line = line;
+        }
+        match status {
+            Status::Success => test.status = TestCaseRunStatus::Success.into(),
+            Status::Failure => test.status = TestCaseRunStatus::Failure.into(),
+            Status::Skipped => test.status = TestCaseRunStatus::Skipped.into(),
+        }
+        // test.status = status;
+        test.attempt_number = attempt_number;
+        let started_at_date_time = DateTime::from_timestamp(started_at, 0).unwrap_or_default();
+        let test_started_at = Timestamp {
+            seconds: started_at_date_time.timestamp(),
+            nanos: started_at_date_time.timestamp_subsec_nanos() as i32,
+        };
+        test.started_at = Some(test_started_at);
+        let finished_at_date_time = DateTime::from_timestamp(finished_at, 0).unwrap_or_default();
+        let test_finished_at = Timestamp {
+            seconds: finished_at_date_time.timestamp(),
+            nanos: finished_at_date_time.timestamp_subsec_nanos() as i32,
+        };
+        test.finished_at = Some(test_finished_at);
+        test.status_output_message = output;
+        self.0
+            .borrow_mut()
+            .test_result
+            .test_case_runs
+            .push(test.clone());
+    }
+
+    // lists the quarantined tests in the test report
+    pub fn list_quarantined_tests(&self) {
+        todo!("List quarantined");
+    }
+
+    // checks if a test is quarantined
+    pub fn is_quarantined(&self, _id: String) {
+        todo!("Is quarantined");
+    }
+
+    // validates the env is set for CI
+    pub fn valid_env(&self) {
+        todo!("Valid env");
+    }
+
+    // validates that we are in a git repo
+    pub fn valid_git(&self) {
+        todo!("Valid git");
+    }
```

Comment:
There's are all to be implemented at a later date. I can remove them for now if we don't want them as is in the repo.